pandas/tests/io/test_orc.py

```diff
@@ -224,3 +225,79 @@ def test_orc_reader_snappy_compressed(dirpath):
     got = read_orc(inputfile).iloc[:10]
 
     tm.assert_equal(expected, got)
+
+
+def test_orc_roundtrip_file(dirpath):
+    # GH44554
+    # PyArrow gained ORC write support with the current argument order
+    pytest.importorskip("pyarrow", minversion="7.0.0")
+    data = {
+        "boolean1": np.array([False, True], dtype="bool"),
+        "byte1": np.array([1, 100], dtype="int8"),
+        "short1": np.array([1024, 2048], dtype="int16"),
+        "int1": np.array([65536, 65536], dtype="int32"),
+        "long1": np.array([9223372036854775807, 9223372036854775807], dtype="int64"),
+        "float1": np.array([1.0, 2.0], dtype="float32"),
+        "double1": np.array([-15.0, -5.0], dtype="float64"),
+        "bytes1": np.array([b"\x00\x01\x02\x03\x04", b""], dtype="object"),
+        "string1": np.array(["hi", "bye"], dtype="object"),
+    }
+    expected = pd.DataFrame.from_dict(data)
+
+    outputfile = os.path.join(dirpath, "TestOrcFile.testReadWrite.orc")
+    expected.to_orc(outputfile)
+    got = read_orc(outputfile)
+
+    tm.assert_equal(expected, got)
+
+
+def test_orc_roundtrip_bytesio():
+    # GH44554
+    # PyArrow gained ORC write support with the current argument order
+    pytest.importorskip("pyarrow", minversion="7.0.0")
+    data = {
+        "boolean1": np.array([False, True], dtype="bool"),
+        "byte1": np.array([1, 100], dtype="int8"),
+        "short1": np.array([1024, 2048], dtype="int16"),
+        "int1": np.array([65536, 65536], dtype="int32"),
+        "long1": np.array([9223372036854775807, 9223372036854775807], dtype="int64"),
+        "float1": np.array([1.0, 2.0], dtype="float32"),
+        "double1": np.array([-15.0, -5.0], dtype="float64"),
+        "bytes1": np.array([b"\x00\x01\x02\x03\x04", b""], dtype="object"),
+        "string1": np.array(["hi", "bye"], dtype="object"),
+    }
+    expected = pd.DataFrame.from_dict(data)
+
+    bytes = expected.to_orc()
+    got = read_orc(BytesIO(bytes))
+
+    tm.assert_equal(expected, got)
+
+
+testdata = [
```

Comment:
can you make this into a fixture or put it at the top. its not obvious what this is used for (yes its in the function )b ut better with a big comment, these are not implemented

Reply:
@jreback  If we are supposed to support categorical columns in 1.5 we need to convert them to some other type such as string/object as the underlying ORC writer for dictionary type is not yet available in PyArrow (see https://issues.apache.org/jira/browse/ARROW-11197 which I can work on ASAP but it won't make it into Arrow before our 1.5.0 release due to the Arrow release schedule).

Everything else has been done.

Reply:
Here is our previous conversation on this issue a week ago:

> > @twoertwein There is one more decision I’d like to consult you about: Currently the ORC writer in pyarrow doesn’t handle dictionary type which correspond to categorical type in pandas.
> > Hence for dataframes with one or more categorical columns we probably need to do one of the following:
> > 
> > 1. Raise ValueError and explain the issue in Arrow.
> > 2. Convert categorical columns into string / object columns before proceeding. Maybe we can warn users about the issue.
> > 
> > On the Arrow side I can get dictionary type enabled in the ORC writer in 9.0.0 if necessary. However given our release schedules it will likely be slightly after the next Pandas release.
> > Since Pandas needs backward compatibility the choice we make here is likely to remain until 2.0.0. Which one do you think is more appropriate? Personally I’m inclined towards 2 since categorical columns are quite common and that it is quite natural to change the behavior once we improve other typing-related aspects of Pandas according to the roadmap such as adding string type and making null handling consistent.
> 
> ok can you add a tests which fails for this case. i wouldn't worry too much here about requiring a higher arrow release, we do this all the time for various types. e.g. when its available in arrow we can add it for the next pandas release, nbd.

Please let me know if we actually need to enable it in 1.5.0.

Reply:
maybe i was not clear

totally fine these are not implemented 

i just want to make it obvious by the test name / fixture name that is the case

eg

test_dtypes_not_supported or similar 

my main issue was that testdata for the fixture name is not obvious - so change the name and add some comments

Reply:
@jreback Really thanks for clarification! Now the test name is `test_orc_writer_dtypes_not_supported` and `testdata` has been renamed `orc_writer_dtypes_not_supported`.